dask/dataframe/multi.py

```diff
@@ -230,7 +230,51 @@ def require(divisions, parts, required=None):
 
 def merge_chunk(lhs, *args, **kwargs):
     empty_index_dtype = kwargs.pop("empty_index_dtype", None)
-    out = lhs.merge(*args, **kwargs)
+
+    rhs = args[0] if len(args) > 0 else None
+    left_index = kwargs.get("left_index", False)
+    right_index = kwargs.get("right_index", False)
+
+    # discover categoricals
+    left_cats = lhs.select_dtypes(include="category")
+    right_cats = rhs.select_dtypes(include="category")
+
+    for col in set(left_cats.columns).union(right_cats.columns):
+        left = None
+        right = None
+
+        if col in left_cats:
+            left = left_cats[col]
+        elif col == kwargs.get("right_on", None) and left_index:
+            left = lhs.index
```

Comment:
Can you double check: is this correct for something like

```python
In [12]: df1 = pd.DataFrame({"A": [0, 1], "B": pd.Categorical(['a', 'b'])})
    ...: df2 = pd.DataFrame({"C": pd.Categorical(['a', 'b'])})
    ...: pd.merge(df2, df1, left_index=True, right_on="A")
Out[12]:
   C  A  B
0  a  0  a
1  b  1  b
```

In other words, does `col` missing from `left_cats` (`C` in this case) and `right_on="A"` and `left_index` necessarily imply that `left = lhs.index`? Or could it just not be in `lhs`?

Reply:
I'll add a test for that, but yeah it works. I think the situation that might not be handled properly is if `"A"` is a categorical but `lhs.index` is not. I think right now it'll be coerced to a categorical.

Reply:
Ok I am handling that now as well. I went back to passing categorical columns in from the meta since, that seemed like the only reliable way to ensure that the dtypes for each partition match.